From f7ac88807e7354faaab881ed56e2ef7f0f11884e Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Tue, 04 May 2021 16:13:42 +0300
Subject: [PATCH] [SW-0] i2c indirect: wd: raise error status

If a reset cause is update in fw_info raise the error status.

Signed-off-by: Noam Lessner <nlessner@habana.ai>
Change-Id: I6eff740ce771bfff922efad5f3694b5c0939437c
---

diff --git a/include/i2cslave_indirect.h b/include/i2cslave_indirect.h
index b57fcd4..6074ccf 100644
--- a/include/i2cslave_indirect.h
+++ b/include/i2cslave_indirect.h
@@ -25,6 +25,7 @@
 bool i2csi_ctrl_is_rst_dev(void);
 void i2csi_set_boot_err(enum hl_fw_fit_boot_error err,
 			bool is_linux_boot_fail);
+int i2csi_set_wd_err(enum hl_fw_reset_cause err);
 void i2csi_update_smbus_alert(struct rma_entry *rma_header);
 extern struct rma_failure i2c_rma_failure;
 #endif /* _I2CSLAVE_INDIRECT_H_ */
diff --git a/preboot/i2cslave_indirect.c b/preboot/i2cslave_indirect.c
index 6e672eb..4e6c375 100644
--- a/preboot/i2cslave_indirect.c
+++ b/preboot/i2cslave_indirect.c
@@ -1097,3 +1097,17 @@
 			sig_fail->uboot_failure = 1;
 	}
 }
+
+int i2csi_set_wd_err(enum hl_fw_reset_cause err)
+{
+	u32 err_sts_wd_err = HL_ERROR_STATUS_OTHER |
+					(HL_ERROR_STATUS_OTHER_WATCHDOG << 8);
+
+	if (err != HL_FW_RESET_CAUSE_UNKNOWN) {
+		fw_info.reset_cause = err;
+		err_sts_reg_val |= err_sts_wd_err;
+		return 0;
+	}
+
+	return 1;
+}
diff --git a/preboot/preboot.c b/preboot/preboot.c
index 4e1cb94..2dfa787 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -396,7 +396,7 @@
 	preboot_process_last_rma_info();
 
 	if (hl_is_watchdog_reset())
-		fw_info.reset_cause = HL_FW_RESET_CAUSE_WATCHDOG;
+		i2csi_set_wd_err(HL_FW_RESET_CAUSE_WATCHDOG);
 
 	module_handle_cmds(cmd_ops, ARRAY_SIZE(cmd_ops));
 }
